re #319 Add capability to TlmGrapher to graph elements of array items…

donaldatball opened the pull request:
….  Also update PacketView and TlmViewer to use a new dialog that detects an attempt to graph an array item and asks the user to specify an array index (or all).

ryanatball commented on lib/cosmos/tools/tlm_grapher/tlm_grapher.rb:
```diff
@@ -74,9 +74,19 @@ def closeEvent(event)
+        packet_name.upcase!
+        item_name.upcase!
+        if (item_name =~ /\[\d+\]$/)
+          item_name = $`
```
Please comment this magic :)

ryanatball commented on lib/cosmos/tools/tlm_grapher/data_object_adders/housekeeping_data_object_adder.rb:
```diff
@@ -64,9 +71,10 @@ def update
       data_object = HousekeepingDataObject.new
       data_object.set_item(target_name, packet_name, item_name)
+      data_object.item_array_index = item_array_index
       @add_data_object_callback.call(data_object) if @add_data_object_callback
```
The adder doesn't work if select an array item without doing [0] etc.  It should work and default to the first item in the array

ryanatball commented on lib/cosmos/tools/tlm_grapher/data_object_adders/housekeeping_data_object_adder.rb:
```diff
@@ -38,14 +38,21 @@ def initialize(parent, orientation = Qt::Horizontal)
           packet_name = split_tlm[1]
-          item_name = split_tlm[2]
+          if (split_tlm[2] =~ /\[\d+\]$/)
+            item_name = $`
```
Please document fancy regex logic and use of $ variables.

donaldatball commented:
@ryanatball Ok, added comments and fixed the adder.  I also added one more line to protect against the array index notation being used for non-array items in the adder.

ryanatball commented:
:+1: Looks good.
